### ISSUE TITLE
source code example install java crash

### ISSUE THREAD

**Issue by daveinman** [fpinscala/fpinscala#555]
**Problem:**
When trying to install the example code for the FP in Scala it fails.

NOTE: issue 548 looks similar, but is closed.  It blames the problem on 'sbt'.  I am running the lastest
sbt I can find for the mac.

I am new to scala and java.  Lots of C++ and ruby in my background.

The only reason I was installing the examples is because code in the book did not run
without a definition of Cons() for work with Lists and I hoped the install would provide
a definition.

**Environment**
My computer is a Mac book pro running Sierra 10.12.6
sbt version is 1.2.8
scala -version
    * Scala code runner version 2.11.4 -- Copyright 2002-2013, LAMP/EPFL
java -version
    * openjdk 11.0.1 2018-10-16
    * OpenJDK Runtime Environment 18.9 (build 11.0.1+13)
    * OpenJDK 64-Bit Server VM 18.9 (build 11.0.1+13, mixed mode)

**Here's a log of the problem:**
```
% rm -rf ~/.sbt
% git clone https://github.com/fpinscala/fpinscala.git
Cloning into 'fpinscala'...
remote: Enumerating objects: 3052, done.
remote: Total 3052 (delta 0), reused 0 (delta 0), pack-reused 3052
Receiving objects: 100% (3052/3052), 5.01 MiB | 7.01 MiB/s, done.
Resolving deltas: 100% (944/944), done.
%  cd fpinscala; ls
LICENSE		answers		project		sbt.bat
README.md	build.sbt	sbt		sbt.cmd
answerkey	exercises	sbt-launch.jar
% chmod +x  ./sbt
% ./sbt
OpenJDK 64-Bit Server VM warning: Ignoring option MaxPermSize; support was removed in 8.0
Getting org.scala-sbt sbt 0.13.13 ...
:: retrieving :: org.scala-sbt#boot-app
	confs: [default]
	50 artifacts copied, 0 already retrieved (17645kB/121ms)
Getting Scala 2.10.6 (for sbt)...
:: retrieving :: org.scala-sbt#boot-scala
	confs: [default]
	5 artifacts copied, 0 already retrieved (24494kB/94ms)
[info] Loading project definition from /Users/davei/fpinscala/project
[info] Updating {file:/Users/davei/fpinscala/project/}fpinscala-build...
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by sbt.ivyint.ErrorMessageAuthenticator$ (file:/Users/davei/.sbt/boot/scala-2.10.6/org.scala-sbt/sbt/0.13.13/ivy-0.13.13.jar) to field java.net.Authenticator.theAuthenticator
WARNING: Please consider reporting this to the maintainers of sbt.ivyint.ErrorMessageAuthenticator$
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[info] Resolving org.scala-sbt.ivy#ivy;2.3.0-sbt-2cf13e211b2cb31f0d3b317289dca70[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
java.lang.NullPointerException
	at java.base/java.util.regex.Matcher.getTextLength(Matcher.java:1770)
	at java.base/java.util.regex.Matcher.reset(Matcher.java:416)
	at java.base/java.util.regex.Matcher.<init>(Matcher.java:253)
	at java.base/java.util.regex.Pattern.matcher(Pattern.java:1133)
	at java.base/java.util.regex.Pattern.split(Pattern.java:1261)
	at java.base/java.util.regex.Pattern.split(Pattern.java:1334)
	at sbt.IO$.pathSplit(IO.scala:797)
	at sbt.IO$.parseClasspath(IO.scala:912)
	at sbt.compiler.CompilerArguments.extClasspath(CompilerArguments.scala:66)
	at sbt.compiler.MixedAnalyzingCompiler$.withBootclasspath(MixedAnalyzingCompiler.scala:188)
	at sbt.compiler.MixedAnalyzingCompiler$.searchClasspathAndLookup(MixedAnalyzingCompiler.scala:166)
	at sbt.compiler.MixedAnalyzingCompiler$.apply(MixedAnalyzingCompiler.scala:176)
	at sbt.compiler.IC$.incrementalCompile(IncrementalCompiler.scala:138)
	at sbt.Compiler$.compile(Compiler.scala:155)
	at sbt.Compiler$.compile(Compiler.scala:141)
	at sbt.Defaults$.sbt$Defaults$$compileIncrementalTaskImpl(Defaults.scala:879)
	at sbt.Defaults$$anonfun$compileIncrementalTask$1.apply(Defaults.scala:870)
	at sbt.Defaults$$anonfun$compileIncrementalTask$1.apply(Defaults.scala:868)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
[error] (compile:compileIncremental) java.lang.NullPointerException
Project loading failed: (r)etry, (q)uit, (l)ast, or (i)gnore? i
[warn] Ignoring load failure: no project loaded.
> 
```

**Comment by omendezmorales** [comment]
Hi @daveinman ,
I'm by no means an expert in sbt  nor Scala, however I was curious and tried to replicate your bug -albeit in two distinct platforms (i.e. Windows 10, and Ubuntu (16.04.5 LTS)). In both  cases I succeded to download and run sbt. What draws my attention from your log above, are the five warnings displayed during the loading of sbt, because I can tell you I did not see those during the same process in the platforms I tried. Maybe it is worth to remove what you have and try once more? Again, I'm neither MacOS user nor Scala expert, but perhaps [this link might help you ](http://sourabhbajaj.com/mac-setup/Scala/README.html)to ensure both sbt & scala are properly installed and configured, prior to cloning this repo. Good luck!

**Comment by daveinman** [comment]
Orlando,

Thanks so much for the time you spent replying to me - it's very thoughtful.  I am away from my scala platform, but will try your suggestion as I get back to it.

Dave

> David M. Inman
> 8968 Southwest Joe Court | Tigard, Oregon 97223 | USA
> Office: +1 (503) 244-1516
> Offshore Milling Services |  www.offshoremilling.com

**Comment by lezhangnet** [comment]
Try to use JDK 8.

**Comment by zored** [comment]
JDK 8 works great. Here is docker command for building:
```
docker run --rm -v $PWD:/app --workdir /app openjdk:8 ./sbt
```

**Comment by mezhaka** [comment]
I have built on top of what @zored suggested to get a persistent container where I run the examples and use sbt.  The difference is that the container persists, otherwise sbt is downloading all the files it needs each time you run it.

Run the following from the root of fpinscala git repo.  This would create a named `fpinscala` container and mount the current directory into the container:
```
docker create -it --name fpinscala -v $PWD:/app --workdir /app openjdk:8
```
Start the container:
```
docker start fpinscala
```
Run bash in the container and attach to it:
```
docker exec -it fpinscala bash
```
Then I was able to follow what README.md suggest, that is, run `./sbt`, select project, and compile.  If you then exit it with Ctrl-D, you can re-attach with 
```
docker container attach fpinscala
```

**Comment by mezhaka** [comment]
I used to get the similar error and ended up getting something different:
```
[ERROR] Failed to construct terminal; falling back to unsupported
java.lang.NumberFormatException: For input string: "0x100"
```
So far I have tried to install [sbt](https://www.scala-sbt.org/download.html) and [JDK](https://adoptopenjdk.net/) recommended on Scala web [site](https://www.scala-lang.org/download/), but I still get lengthy error messages starting with

For sbt I did:
```
echo "deb https://dl.bintray.com/sbt/debian /" | sudo tee -a /etc/apt/sources.list.d/sbt.list
curl -sL "https://keyserver.ubuntu.com/pks/lookup?op=get&search=0x2EE0EA64E40A89B84B2DF73499E82A75642AC823" | sudo apt-key add
sudo apt-get update
sudo apt-get install sbt
```

For JDK:
```
wget -qO - https://adoptopenjdk.jfrog.io/adoptopenjdk/api/gpg/key/public | sudo apt-key add -
sudo add-apt-repository --yes https://adoptopenjdk.jfrog.io/adoptopenjdk/deb/
sudo apt-get update
sudo apt-get install adoptopenjdk-8-hotspot
```

Now I have
```$sbt --version
sbt version in this project: 1.3.13
sbt script version: 1.3.13
$sbt --version
sbt version in this project: 1.3.13
sbt script version: 1.3.13
```

I am able to follow the basic tutorial here https://docs.scala-lang.org/getting-started/sbt-track/getting-started-with-scala-and-sbt-on-the-command-line.html, but when I run `./sbt` within the `fpinscala` directory I still get the error.

**Comment by stevanmilic** [comment]
This https://github.com/fpinscala/fpinscala/pull/589 PR solves the issue with Java 11 to make it compile successfully. You can copy the changes to make it work.

@mezhaka you got `java.lang.NumberFormatException: For input string: "0x100"` because you need to add `export TERM=xterm-color` before you run sbt. Check this [answer](https://stackoverflow.com/a/44361749/2413731) for more info. But that's the case for older version of `sbt` (or so I think).

**Comment by frepettov0** [comment]
In the SO [link](https://stackoverflow.com/a/44361749/2413731) posted by @stevanmilic there is a simpler solution which is changing the sbt version to at least `1.13.16`. Edit the `project/build.properties`
```
sbt.version=0.13.16
```
Compile on a terminal

**Comment by mpilquist** [comment]
Fixed as of 7c0cf46d9b9f8b41a27c1b46872cdaf76febd7d7
